app/controllers/errors_controller.rb

```diff
@@ -0,0 +1,5 @@
+class ErrorsController < ApplicationController
+  def not_found
+    render file: "#{Rails.root}/public/404.html", status: :not_found
```

Comment:
`Rails.root` returns an instance of [Pathname](https://ruby-doc.org/stdlib-3.1.0/libdoc/pathname/rdoc/Pathname.html). This object has the method `join` can receive multiple strings as parameters that will compose a new Pathname object. The advantage of using Pathname instead of string interpolation as you've done is that Pathname will concatenate the strings using `File::SEPARATOR` instead of hardcoded `/`, which makes your code more flexible to work in different Operational Systems (windows or UNIX derivatives)

```suggestion
    render file: Rails.root.join("public", "404.html"), status: :not_found
```